pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,212 @@
+ * Code paths that do no need to rebuild the index or which don't happen on the server need not utilize this throttler.
+ */
+public class SegmentPreprocessThrottler implements PinotClusterConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessThrottler.class);
+
+  /**
+   * _maxPreprocessConcurrency and _maxConcurrentPreprocessesBeforeServingQueries must be >= 0. To effectively disable
+   * throttling, this can be set to a very high value
+   */
+  private int _maxPreprocessConcurrency;
+  private int _maxPreprocessConcurrencyBeforeServingQueries;
+  private boolean _relaxThrottling;
+  private final AdjustableSemaphore _semaphore;
+
+  /**
+   * @param maxPreprocessConcurrency configured preprocessing concurrency
+   * @param maxPreprocessConcurrencyBeforeServingQueries configured preprocessing concurrency before serving queries
+   * @param relaxThrottling whether to relax throttling prior to serving queries
+   */
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, int maxPreprocessConcurrencyBeforeServingQueries,
+      boolean relaxThrottling) {
+    LOGGER.info("Initializing SegmentPreprocessThrottler, maxPreprocessConcurrency: {}, "
+            + "maxPreprocessConcurrencyBeforeServingQueries: {}, relaxThrottling: {}",
+        maxPreprocessConcurrency, maxPreprocessConcurrencyBeforeServingQueries, relaxThrottling);
+    Preconditions.checkArgument(maxPreprocessConcurrency > 0,
+        "Max preprocess parallelism must be > 0, but found to be: " + maxPreprocessConcurrency);
+    Preconditions.checkArgument(maxPreprocessConcurrencyBeforeServingQueries > 0,
+        "Max preprocess parallelism before serving queries must be > 0, but found to be: "
+            + maxPreprocessConcurrencyBeforeServingQueries);
+
+    _maxPreprocessConcurrency = maxPreprocessConcurrency;
+    _maxPreprocessConcurrencyBeforeServingQueries = maxPreprocessConcurrencyBeforeServingQueries;
+    _relaxThrottling = relaxThrottling;
+
+    // maxConcurrentPreprocessesBeforeServingQueries is only used prior to serving queries and once the server is
+    // ready to serve queries this is not used again. Thus, it is safe to only pick up this configuration during
+    // server startup. There is no need to allow updates to this via the ZK CLUSTER config handler
+    int relaxThrottlingThreshold = Math.max(_maxPreprocessConcurrency, _maxPreprocessConcurrencyBeforeServingQueries);
+    int preprocessConcurrency = _maxPreprocessConcurrency;
+    if (relaxThrottling) {
+      preprocessConcurrency = relaxThrottlingThreshold;
+      LOGGER.info("Relax throttling enabled, setting preprocess concurrency to: {}", preprocessConcurrency);
+    }
+    _semaphore = new AdjustableSemaphore(preprocessConcurrency, true);
+    LOGGER.info("Created semaphore with total permits: {}, available permits: {}", totalPermits(),
+        availablePermits());
+  }
+
+  public synchronized void resetThrottling() {
```

Comment:
Consider changing the method name to `startServingQueries` to be more specific

Reply:
done